### ISSUE TITLE
Support Mixed OS Version

### ISSUE THREAD

**Issue by AStrangwood** [NVIDIA/gpu-operator#328]
### 1. Issue or feature description

I have a hand full of GPU nodes that have to be on Ubuntu 18.04. I also have a bunch more GPU nodes that should have Ubuntu 20.04 on them like the rest of the cluster nodes however, the GPU Operator cant handle the mixed OS version.  Having the mixed OS versions makes the GPU Operator a paperweight as it will break the working nvidia driver installs on the GPU nodes that were already happily set up and running. Steps below of what exactly happens. 

This is extremely frustrating as i have to deploy Ubuntu 18.04 on all the GPU nodes to be able to use the GPU Operator it also means for now i have no upgrade path away from Ubuntu 18.04 at this time.
 

### 2. Steps to reproduce the issue

1. Install GPU Operator
2. GPU nodes on Ubuntu 18.04 already in the cluster.
3. Wait for drivers to be installed and ready
4. Add GPU node that's on Ubuntu 20.04
5. GPU Operator tries to install the Ubuntu 20.04 driver on this new node. It also unloads the driver from the other GPU nodes and tries to install the ubuntu 20.04 driver. This will fail at some point and then the Operator unloads the driver on every single GPU node and tries to install the Ubuntu 18.04 driver, again this fails and it repeats unloading the driver on all nodes and trys the ubuntu 20.04 driver. This happens in an infinite loop until you remove the mixed Ubuntu OS nodes

**Comment by shivamerla** [comment]
@AStrangwood yes, this is a known limitation with GPU operator today and we are looking to support mixed mode soon.

**Comment by k8s-gpubuilder** [comment]
What about if you don't use the driver through the gpu-operator and install it on the OS instead?

**Comment by shivamerla** [comment]
@k8s-gpubuilder you can pass `--set driver.enabled=false` during operator install. Same for toolkit if its pre-installed on the node.

**Comment by k8s-gpubuilder** [comment]
So is it then supporting several different OS versions?

**Comment by shivamerla** [comment]
yes, driver is the only component OS specific, rest all will work across ubuntu18.04 or ubuntu20.04. Not for mixed OS types though as container-toolkit is different for CentOS vs Ubuntu.

**Comment by bandersen23** [comment]
Is there a future where gpu operator supports multiple OS in the case of both Ubuntu and CentOS?

**Comment by age9990** [comment]
@shivamerla Is there any progress about this feature? There's no way for us to upgrade OS version with this constraint.

**Comment by bandersen23** [comment]
@shivamerla this would be a great feature to add. It would be useful as we have nodes running DGX OS and RHEL.

**Comment by cdesiniotis** [comment]
Pre-installing the NVIDIA drivers on your worker nodes is one way to overcome this limitation. Another approach is to use the NVIDIA Driver CRD for managing GPU driver daemonsets: https://docs.nvidia.com/datacenter/cloud-native/gpu-operator/latest/gpu-driver-configuration.html With this CRD, it is possible to deploy separate GPU driver daemonsets on different pools of nodes (where each pool could be running a different OS version).
